File: cpp/testsuite/ApplicationCases/SceneEditorApplication/src/model/history/SceneTransformationOperation.cpp
#include "java/lang/System.h"
#include "java/util/ArrayList.txx"
#include "java/util/Collections.h"
#include "model/history/EntityTransformState.h"
#include "model/history/SceneTransformationOperation.h"

namespace {
void copyStates(const java::ArrayList<EntityTransformState*>& source,
                java::ArrayList<EntityTransformState*>& target)
{
    long i;
    for ( i = 0; i < source.size(); i++ ) {
        target.add(source.get(i)->clone());
    }
}

void deleteStates(java::ArrayList<EntityTransformState*>& states)
{
    long i;
    for ( i = 0; i < states.size(); i++ ) {
        delete states.get(i);
    }
    states.clear();
}
}

SceneTransformationOperation::SceneTransformationOperation(
    const java::String& name,
    const java::ArrayList<EntityTransformState*>& before,
    const java::ArrayList<EntityTransformState*>& after,
    bool mergeable)
    : name(name), mergeable(mergeable),
      lastChangeTime(java::System::currentTimeMillis())
{
    copyStates(before, this->before);
    copyStates(after, this->after);
}

SceneTransformationOperation::~SceneTransformationOperation()
{
    deleteStates(before);
    deleteStates(after);
}

void SceneTransformationOperation::undo()
{
    long i;
    for ( i = 0; i < before.size(); i++ ) {
        before.get(i)->restore();
    }
}

void SceneTransformationOperation::redo()
{
    long i;
    for ( i = 0; i < after.size(); i++ ) {
        after.get(i)->restore();
    }
}

java::String SceneTransformationOperation::getName() const
{
    return name;
}

int SceneTransformationOperation::getEntityCount() const
{
    return (int)before.size();
}

bool SceneTransformationOperation::absorb(UndoableOperation* next)
{
    SceneTransformationOperation* other =
        dynamic_cast<SceneTransformationOperation*>(next);

    if ( other == nullptr ||
         !mergeable || !other->mergeable || !name.equals(other->name) ||
         other->lastChangeTime - lastChangeTime > MERGE_INTERVAL_MILLISECONDS ||
         !sameElements(other) ) {
        return false;
    }
    // Takes the placements after the other operation; the old ones go to
    // the other operation, deleted by the caller
    java::Collections::swap(after, other->after);
    lastChangeTime = other->lastChangeTime;
    return true;
}

bool SceneTransformationOperation::sameElements(
    const SceneTransformationOperation* other) const
{
    long i;

    if ( other->before.size() != before.size() ) {
        return false;
    }
    for ( i = 0; i < before.size(); i++ ) {
        if ( other->before.get(i)->getEntity() !=
             before.get(i)->getEntity() ) {
            return false;
        }
    }
    return true;
}

File: java/testsuite/ApplicationCases/SceneEditorApplication/src/model/history/SceneTransformationOperation.java
package model.history;

import java.util.ArrayList;
import java.util.List;

/**
Change of the placement (translation, rotation, scale) of some elements of a
scene: bodies, lights or cameras. It keeps the placements before and after
the change of each changed element.

Operations done by the keyboard over the same elements in a short time can
be merged (see `absorb`), so an auto-repeated key is a single step.
*/
public class SceneTransformationOperation implements UndoableOperation
{
    /// Maximum time between two mergeable operations to be merged
    public static final long MERGE_INTERVAL_MILLISECONDS = 1000;

    private final String name;
    private final ArrayList<EntityTransformState> before;
    private ArrayList<EntityTransformState> after;
    private final boolean mergeable;
    private long lastChangeTime;

    /**
    @param name name of the user action
    @param before placements before the change, one per changed element
    @param after placements after the change, in the same order
    @param mergeable true if following operations over the same elements
    can be merged into this one
    */
    public SceneTransformationOperation(String name,
                                        List<EntityTransformState> before,
                                        List<EntityTransformState> after,
                                        boolean mergeable)
    {
        this.name = name;
        this.before = new ArrayList<EntityTransformState>(before);
        this.after = new ArrayList<EntityTransformState>(after);
        this.mergeable = mergeable;
        this.lastChangeTime = System.currentTimeMillis();
    }

    @Override
    public void undo()
    {
        for ( EntityTransformState state : before ) {
            state.restore();
        }
    }

    @Override
    public void redo()
    {
        for ( EntityTransformState state : after ) {
            state.restore();
        }
    }

    @Override
    public String getName()
    {
        return name;
    }

    /**
    @return number of entities whose placement changed
    */
    public int getEntityCount()
    {
        return before.size();
    }

    @Override
    public boolean absorb(UndoableOperation next)
    {
        if ( !(next instanceof SceneTransformationOperation other) ||
             !mergeable || !other.mergeable || !name.equals(other.name) ||
             other.lastChangeTime - lastChangeTime > MERGE_INTERVAL_MILLISECONDS ||
             !sameElements(other) ) {
            return false;
        }
        after = other.after;
        lastChangeTime = other.lastChangeTime;
        return true;
    }

    private boolean sameElements(SceneTransformationOperation other)
    {
        int i;

        if ( other.before.size() != before.size() ) {
            return false;
        }
        for ( i = 0; i < before.size(); i++ ) {
            if ( other.before.get(i).getEntity() != before.get(i).getEntity() ) {
                return false;
            }
        }
        return true;
    }
}
